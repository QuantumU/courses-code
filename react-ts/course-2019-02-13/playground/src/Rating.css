.star-active {
  color: gold;
}

.star-inactive {
  color: lightgrey;
}