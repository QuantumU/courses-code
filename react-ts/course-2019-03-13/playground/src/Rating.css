.star {
  font-size: 3rem;
  color: lightgrey;
  user-select: none;
}

.star.active {
  color: gold;
}
