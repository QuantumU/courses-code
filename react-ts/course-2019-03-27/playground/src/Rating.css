.star-inactive {
  color: lightgrey;
  font-size: 40px;
}

.star-active {
  color: gold;
  font-size: 40px;
}