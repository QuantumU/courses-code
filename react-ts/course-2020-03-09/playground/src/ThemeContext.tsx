import { createContext } from "react";

export type ThemeName = "default" | "fancy";

const ThemeContext = createContext<ThemeName>("default");

export default ThemeContext;
