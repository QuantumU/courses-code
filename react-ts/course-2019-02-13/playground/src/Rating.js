import React, { Component } from 'react';
import './Rating.css';

export default class Rating extends Component {
  /*
    props:
      stars: number
    events:
      onStarsChange: function(number)
  */
  render() {
    const starData = [];
    for (let i = 1; i <= 5; i++) {
      starData.push({
        id: i,
        active: i <= this.props.stars
      });
    }
    return (
      <div>
        {starData.map(star => (
          <span
            key={star.id}
            className={star.active ? 'star-active' : 'star-inactive'}
            onClick={() => {this.props.onStarsChange(star.id)}}
          >
            {star.active ? '★' : '☆'}
          </span>
        ))}
      </div>
    );
  }
}
