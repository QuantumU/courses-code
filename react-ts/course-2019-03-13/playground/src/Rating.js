import React, { Component } from 'react';

import './Rating.css';

class Rating extends Component {
  // props:
  //   stars: number
  // events:
  //   onStarsChange: (stars: number) => void

  shouldComponentUpdate(nextProps, nextState) {
    return this.props.stars !== nextProps.stars;
  }

  render() {
    const starIndexes = [1, 2, 3, 4, 5];

    return (
      <div>
        {starIndexes.map(index =>
          this.props.stars < index ? (
            <span
              className="star"
              key={index}
              onClick={() => {
                this.props.onStarsChange(index);
              }}
            >
              ☆
            </span>
          ) : (
            <span
              className="star active"
              key={index}
              onClick={() => {
                this.props.onStarsChange(index);
              }}
            >
              ★
            </span>
          )
        )}
      </div>
    );
  }
}

export default Rating;
