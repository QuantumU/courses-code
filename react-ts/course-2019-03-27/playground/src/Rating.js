import React from 'react';
import './Rating.css';
import {range} from 'lodash'

const Rating = React.memo(props => {
  /**
   * props:
   *   stars: number
   * event:
   *   onStarsChange: (newStars: number) => undefined
   */
  const starIndexes = range(1, 6);

  return starIndexes.map(index => (
    <span
      className={index <= props.stars ? 'star-active' : 'star-inactive'}
      onClick={() => {
        props.onStarsChange(index);
      }}
      key={index}
    >
      {index <= props.stars ? '★' : '☆'}
    </span>
  ));
});

export default Rating;
