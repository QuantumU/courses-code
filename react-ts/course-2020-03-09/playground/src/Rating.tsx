import React, { useContext } from "react";
import ThemeContext from "./ThemeContext";

type Props = {
  stars: number;
  onStarsChange?: (newStars: number) => void;
};

function Rating(props: Props) {
  const themeContext = useContext(ThemeContext);

  if (props.stars < 1) {
    throw new Error("number of stars must be 1-5");
  }

  const starElements = [];

  for (let i = 0; i <= 4; i++) {
    const isActive = i + 1 <= props.stars;
    let starSymbol;
    if (themeContext === "default") {
      starSymbol = isActive ? "â˜…" : "â˜†";
    } else {
      starSymbol = isActive ? "ðŸ˜€" : "ðŸ˜¡";
    }
    const starStyle = { color: isActive ? "gold" : "lightgrey" };
    const clickHandler = () => {
      // trigger onStarsChange (if it is defined)
      if (props.onStarsChange !== undefined) {
        props.onStarsChange(i + 1);
      }
    };
    starElements.push(
      <span key={i} style={starStyle} onClick={clickHandler}>
        {starSymbol}
      </span>
    );
  }

  return <div>{starElements}</div>;
}

export default Rating;
