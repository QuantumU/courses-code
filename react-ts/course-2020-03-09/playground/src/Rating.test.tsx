import React from "react";
import { render, fireEvent } from "@testing-library/react";

import Rating from "./Rating";
import ThemeContext, { ThemeName } from "./ThemeContext";

it("renders three full stars", () => {
  const instance = render(
    <ThemeContext.Provider value="default">
      <Rating stars={3} />
    </ThemeContext.Provider>
  );
  const fullStars = instance.getAllByText("★");
  expect(fullStars).toHaveLength(3);
  for (let fullStar of fullStars) {
    expect(fullStar).toHaveStyle("color: gold;");
  }
});

it("triggers an event when the fourth star is clicked", () => {
  const mockFn = jest.fn();
  const instance = render(<Rating stars={3} onStarsChange={mockFn} />);
  const firstEmptyStar = instance.getAllByText("☆")[0];
  fireEvent.click(firstEmptyStar);
  expect(mockFn).toHaveBeenCalledWith(4);
});

it("throws an error if the number of stars is 0", () => {
  const testFn = () => {
    render(<Rating stars={0} />);
  };
  expect(testFn).toThrow("number of stars must be 1-5");
});

it("matches snapshot", () => {
  const instance = render(
    <ThemeContext.Provider value="default">
      <Rating stars={3} />
    </ThemeContext.Provider>
  );
  expect(instance.baseElement).toMatchSnapshot();
});
