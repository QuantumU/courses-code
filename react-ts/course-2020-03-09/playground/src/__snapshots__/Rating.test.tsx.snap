// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<body>
  <div>
    <div>
      <span
        style="color: gold;"
      >
        ★
      </span>
      <span
        style="color: gold;"
      >
        ★
      </span>
      <span
        style="color: gold;"
      >
        ★
      </span>
      <span
        style="color: lightgrey;"
      >
        ☆
      </span>
      <span
        style="color: lightgrey;"
      >
        ☆
      </span>
    </div>
  </div>
</body>
`;
